### PR TITLE
Enable using __setitem__ for Dataset using a list as key

### LINKED ISSUE #1123: Assigning several variables at once in a dataset
Hi,

I cannot tell if I'm doing something illegitimate. I have a hard time to understand the error I'm facing:

```python
import numpy as np
import xarray as xr

ds = xr.Dataset(
    {'varA': (['x', 'time'], np.random.rand(2, 2)),
     'varB': (['x', 'time'], np.random.rand(2, 2)),
     'varC': (['x', 'time'], np.random.rand(2, 2))},
    coords={'x': [0, 1.], 'time': [0, 1.]})

print(ds)

# I want to center 'varA' and 'varB' by substracting their mean values along the time dimension:
ds[['varA', 'varB']] -= ds[['varA', 'varB']].mean(dim='time')
```

Produces the output:
```
<xarray.Dataset>
Dimensions:  (time: 2, x: 2)
Coordinates:
  * x        (x) float64 0.0 1.0
  * time     (time) float64 0.0 1.0
Data variables:
    varA     (x, time) float64 0.7737 0.2608 0.009832 0.8498
    varC     (x, time) float64 0.9732 0.6492 0.5118 0.5792
    varB     (x, time) float64 0.04973 0.4012 0.01248 0.8716
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-36-aa87123fb25d> in <module>()
      7 print(ds)
      8 
----> 9 ds[['varA', 'varB']] -= ds[['varA', 'varB']].mean(dim='time')

C:\Users\poimat04\AppData\Local\Continuum\Miniconda2\envs\AxialLLRCA\lib\site-packages\xarray\core\dataset.pyc in __setitem__(self, key, value)
    536             raise NotImplementedError('cannot yet use a dictionary as a key '
    537                                       'to set Dataset values')
--> 538         self.update({key: value})
    539 
    540     def __delitem__(self, key):

TypeError: unhashable type: 'list'
```

Is it forbidden to assign several variables at once? I'm puzzled with this TypeError.

Thanks

### PR DESCRIPTION
**Pull request by gcaria**
- [x] Closes #1123
- [x] Tests added
- [x] Passes `pre-commit run --all-files`
- [x] User visible changes (including notable bug fixes) are documented in `whats-new.rst`

Hi xarray folks, long time user first time contributor here.

I believe the tests for this feature should be expanded, so please consider this as a work in progress.
Any feedback is greatly appreciated!

### CODE REVIEW

File: xarray/core/dataset.py [hunk 1]
```diff
@@ -1491,7 +1491,15 @@ def __setitem__(self, key: Hashable, value) -> None:
                 "cannot yet use a dictionary as a key to set Dataset values"
             )
 
-        self.update({key: value})
+        if isinstance(key, list):
+            if len(key) != len(value):
+                raise ValueError("keys and value must have the same length.")
```
Comment (max-sixty) [inline xarray/core/dataset.py:1496]: Could we be slightly more descriptive — maybe give the length of each, if not the full values? I'm not sure `key` and `value` alone will be that clear b/c they're not named arguments.

File: xarray/core/dataset.py [hunk 2]
```diff
@@ -1475,14 +1475,14 @@ def __getitem__(self, key):
         else:
             return self._copy_listed(np.asarray(key))
 
-    def __setitem__(self, key: Hashable, value) -> None:
-        """Add an array to this dataset.
+    def __setitem__(self, key, value) -> None:
```
Comment (max-sixty) [inline xarray/core/dataset.py:1478]: Can we add to the typing rather than removing it? Let me know if you want a hand

File: xarray/core/dataset.py [hunk 3]
```diff
@@ -1475,14 +1475,14 @@ def __getitem__(self, key):
...
+        """Add a (list of) array to this dataset.
```
Comment (max-sixty) [inline xarray/core/dataset.py:1479]: I think this could be clearer — check out other methods that take a single item or a list

File: xarray/core/dataset.py [hunk 4]
```diff
@@ -1491,7 +1491,15 @@ def __setitem__(self, key: Hashable, value) -> None:
                 "cannot yet use a dictionary as a key to set Dataset values"
             )
 
-        self.update({key: value})
+        if isinstance(key, list):
+            if len(key) != len(value):
+                raise ValueError("keys and value must have the same length.")
+            if isinstance(value, Dataset):
+                self.update(dict(zip(key, value.data_vars.values())))
```
Comment (max-sixty) [inline xarray/core/dataset.py:1498]: Is this correct? If the `key` and `data_vars` are out of order, is it going to be doing the right thing? It may be OK to exclude this option for the moment.

File: xarray/tests/test_dataset.py [hunk 5]
```diff
@@ -3495,6 +3495,25 @@ def test_setitem_str_dtype(self, dtype):
...
+    def test_setitem_using_list(self):
+
+        # assign a list of variables
+        var1 = Variable(["dim1"], np.random.randn(8))
+        var2 = Variable(["dim1"], np.random.randn(8))
+        data1 = create_test_data()
+        data1[["A", "B"]] = [var1, var2]
+        data2 = data1.copy()
+        data2[["A", "B"]] = [var1, var2]
+        assert_identical(data1, data2)
```
Comment (max-sixty) [inline xarray/tests/test_dataset.py:3507]: I would have a test that adds variables with the existing approach — one at a time — and then compares that result to adding them in the proposed approach — with a list of keys.

**Comment by max-sixty** [comment]
I think this is a good feature — any thoughts from others? Should a dataset be a valid rvalue?

I added some comments @gcaria . Great to have you as a contributor. We could add a line or two in the docs, and whatsnew. And you're right — the tests could be filled out a bit (no need to go overboard though).

File: xarray/core/dataset.py (hunk 2 above)
Comment (gcaria) [reply xarray/core/dataset.py:1486]: Ok this should look better now :)

File: xarray/core/dataset.py (hunk 4 above)
Comment (gcaria) [reply xarray/core/dataset.py:1510]: For the case:
```python
ds[['foo', 'bar']] = another_ds[['bar', 'foo']]
```
it could be safe to assume the user wants to swap variable names since this is quite explicit. Technically there could be a problem with:
```python
ds[['foo', 'bar']] = another_ds 
```
where `another_ds` contains only the variables `foo` and `bar` but in the wrong order. Maybe such cases can be identified and a warning issued (which reminds the user it's better to be explicit as in the first example)

File: xarray/tests/test_dataset.py (hunk 5 above)
Comment (gcaria) [reply xarray/tests/test_dataset.py:3508]: Ok this should look better now :)

**Review by gcaria — COMMENTED** [review]
Thanks for the prompt feedback! I have addressed the comments. By the way, I believe the proposed changes (using a Dataset as valid rvalue) do close #1123 :)

File: xarray/core/dataset.py (hunk 1 above)
Comment (gcaria) [reply xarray/core/dataset.py]: I've tried to be more descriptive and to give a more user-friendly message, although it isn't easy to be concise

File: xarray/core/dataset.py (hunk 3 above)
Comment (gcaria) [reply xarray/core/dataset.py]: Ok this should look better now :)

File: xarray/core/dataset.py (hunk 4 above)
Comment (max-sixty) [reply xarray/core/dataset.py:1512]: Good point — even in the first case it's still a dataset, rather than a list of variables.

It does seem a bit dangerous but also that it's not _that_ easy to do accidentally.

Any thoughts from others?

File: xarray/tests/test_dataset.py (hunk 5 above)
Comment (max-sixty) [reply xarray/tests/test_dataset.py:3508]: Perfect

File: xarray/core/dataset.py (hunk 2 above)
Comment (max-sixty) [reply xarray/core/dataset.py:1486]: Great. I _think_ this could have been a Union, but it's completely fine like this too.

**Comment by max-sixty** [comment]
Great @gcaria ! And thanks for the issue link.

Unless we hear from others, we can merge in the next couple of days. Could we add a test for the ds rvalue and a whatsnew?

File: xarray/core/dataset.py [hunk 6]
```diff
@@ -1491,7 +1501,19 @@ def __setitem__(self, key: Hashable, value) -> None:
                 "cannot yet use a dictionary as a key to set Dataset values"
             )
 
-        self.update({key: value})
+        if isinstance(key, list):
```
Comment (mathause) [inline xarray/core/dataset.py:1504]: I don't think you can use `Iterable` here because `isinstance(xr.DataArray(), Iterable)` is `True`.

File: xarray/core/dataset.py [hunk 7]
```diff
@@ -1491,7 +1501,19 @@ def __setitem__(self, key: Hashable, value) -> None:
...
+            if len(key) != len(value):
+                raise ValueError(
+                    f"Different lengths of variables to be set "
+                    f"({len(key)}) and data used as input for "
+                    f"setting ({len(value)})"
+                )
+            if isinstance(value, Dataset):
+                self.update(dict(zip(key, value.data_vars.values())))
+            else:
+                self.update(dict(zip(key, value)))
```
Comment (mathause) [inline xarray/core/dataset.py:1514]: I think the case `ds[["a"]] = xr.DataArray([1, 2]))` is unlikely but dangerous here. Maybe:

```suggestion
                if isinstance(value, DataArray):
                    raise ValueError("Cannot assign single DataArray to multiple keys")
                self.update(dict(zip(key, value)))
```

File: xarray/tests/test_dataset.py [hunk 8]
```diff
@@ -3495,6 +3495,26 @@ def test_setitem_str_dtype(self, dtype):
...
+
+        # assign a list of variables
+        var1 = Variable(["dim1"], np.random.randn(8))
+        var2 = Variable(["dim1"], np.random.randn(8))
+        data1 = create_test_data()
+        data2 = data1.copy()
+        data1["A"] = var1
+        data1["B"] = var2
+        data2[["A", "B"]] = [var1, var2]
+        assert_identical(data1, data2)
```
Comment (mathause) [inline xarray/tests/test_dataset.py:3508]:
```suggestion
        actual = create_test_data()
        expected = actual.copy()
        expected["A"] = var1
        expected["B"] = var2
        actual[["A", "B"]] = [var1, var2]
        assert_identical(actual, expected)
```
nit

File: xarray/tests/test_dataset.py [hunk 9]
```diff
@@ -3495,6 +3495,26 @@ def test_setitem_str_dtype(self, dtype):
...
+        # assign a list of dataset arrays
+        dv = 2 * data2[["A", "B"]]
+        data1[["C", "D"]] = [d.variable for d in dv.data_vars.values()]
+        data2[["C", "D"]] = dv
+        assert_identical(data1, data2)
```
Comment (mathause) [inline xarray/tests/test_dataset.py:3513]:
```suggestion
        # assign a list of numpy arrays
        actual[["C", "D"]] = [var1.values, var2.values]
        expected[["C", "D"]] = expected[["A", "B"]]
        assert_identical(actual, expected)
```

File: xarray/tests/test_dataset.py [hunk 10]
```diff
@@ -3495,6 +3495,26 @@ def test_setitem_str_dtype(self, dtype):
...
+
+        with pytest.raises(ValueError, match=r"Different lengths"):
+            data1[["A", "B"]] = [var1]
```
Comment (mathause) [inline xarray/tests/test_dataset.py:3516]:
```suggestion
            actual[["A", "B"]] = [var1]
```

File: xarray/core/dataset.py [hunk 11]
```diff
@@ -1475,14 +1475,24 @@ def __getitem__(self, key):
         else:
             return self._copy_listed(np.asarray(key))
 
+    @overload
+    def __setitem__(self, key: Iterable[Hashable], value) -> None:
```
Comment (mathause) [inline xarray/core/dataset.py:1479]:
```suggestion
    def __setitem__(self, key: List[Hashable], value) -> None:
```
You do a `isinstance(key, list)` check below - so it has to be `List` here.

**Review by mathause — COMMENTED** [review]
Looks good  - I found one (maybe) dangerous edge-case and added some nits.

File: xarray/core/dataset.py (hunk 7 above)
Comment (max-sixty) [reply xarray/core/dataset.py:1514]: Nice, good spot @mathause

**Comment by max-sixty** [comment]
Great — @gcaria if you're OK to hit the "Commit suggestion" on those then we can merge?

Thanks for the review @mathause

File: xarray/core/dataset.py (hunk 7 above)
Comment (gcaria) [reply xarray/core/dataset.py]: I'm guessing you mean `ds[["a","b"]] = xr.DataArray([1, 2]))` since the error message is about multiple keys :)

I have added some logic that makes `ds[["a"]] = xr.DataArray([1, 2]))` possible (unless I'm missing some obvious problem with this).

**Review by mathause — COMMENTED** [review]
Looks good - thanks for the update. Somehow the tests did not run - can you merge master and push again? Let me know if you need help or if I should do it.

**Comment by keewis** [comment]
github had trouble with people abusing the "run actions on PRs" feature (namely, it was used to mine cryptocurrencies) so they now require a maintainer to approve every new commit on a first time contributor's PR (for maintainers there's a "Approve and run" button directly above the merge button)

**Comment by mathause** [comment]
thanks for the info, hadn't seen that one.

File: xarray/tests/test_dataset.py [hunk 12]
```diff
@@ -3495,6 +3495,31 @@ def test_setitem_str_dtype(self, dtype):
...
+        assert_identical(actual, expected)
+
+        with pytest.raises(ValueError, match=r"Different lengths"):
+            actual[["A", "B"]] = [var1]
+
+        with pytest.raises(ValueError, match=r"Empty list of variables"):
+            actual[[]] = [var1]
+
+        with pytest.raises(ValueError, match=r"assign single DataArray"):
+            actual[["A", "B"]] = xr.DataArray([1, 2])
```
Comment (andersy005) [inline xarray/tests/test_dataset.py:3521]: Could we put this in a separate, parametrized test? I imagine something along these lines:

```diff
diff --git a/xarray/tests/test_dataset.py b/xarray/tests/test_dataset.py
index d00e6583..19849ffc 100644
--- a/xarray/tests/test_dataset.py
+++ b/xarray/tests/test_dataset.py
@@ -3511,14 +3511,22 @@ class TestDataset:
         expected[["C", "D"]] = expected[["A", "B"]]
         assert_identical(actual, expected)
 
-        with pytest.raises(ValueError, match=r"Different lengths"):
-            actual[["A", "B"]] = [var1]
-
-        with pytest.raises(ValueError, match=r"Empty list of variables"):
-            actual[[]] = [var1]
-
-        with pytest.raises(ValueError, match=r"assign single DataArray"):
-            actual[["A", "B"]] = xr.DataArray([1, 2])
+    @pytest.mark.parametrize(
+        "var_list, data, error_regex",
+        [
+            (
+                ["A", "B"],
+                [Variable(["dim1"], np.random.randn(8))],
+                r"Different lengths",
+            ),
+            ([], [Variable(["dim1"], np.random.randn(8))], r"Empty list of variables"),
+            (["A", "B"], xr.DataArray([1, 2]), r"assign single DataArray"),
+        ],
+    )
+    def test_setitem_using_list_errors(self, var_list, data, error_regex):
+        actual = create_test_data()
+        with pytest.raises(ValueError, match=error_regex):
+            actual[var_list] = data
 
     def test_assign(self):
         ds = Dataset()
```

File: xarray/core/dataset.py (hunk 2 above)
Comment (gcaria) [reply xarray/core/dataset.py:1486]: Tried this and `mypy` complains that `Single overload definition, multiple required`, it seems related to this https://github.com/python/mypy/issues/5047

**Comment by gcaria** [comment]
Thanks all for the feedback !
This thread it's getting a bit hard to follow, but I believe this is the only suggested change that I haven't applied https://github.com/pydata/xarray/pull/5216#discussion_r624062989

File: xarray/core/dataset.py [hunk 13]
```diff
@@ -1475,14 +1475,24 @@ def __getitem__(self, key):
...
+        ...
+
+    @overload
     def __setitem__(self, key: Hashable, value) -> None:
+        ...
+
+    def __setitem__(self, key, value) -> None:
         """Add an array to this dataset.
+        Multiple arrays can be added at the same time, in which case each of
+        the following operations is applied to the relevant type.
```
Comment (max-sixty) [inline xarray/core/dataset.py:1489]:
```suggestion
        the following operations is applied to the respective value.
```

File: xarray/tests/test_dataset.py (hunk 9 above)
Comment (max-sixty) [reply xarray/tests/test_dataset.py:3513]: Good point — it doesn't align on dim. I'm getting:

```python
AssertionError: Left and right Dataset objects are not identical
Differing dimensions:
    (C: 8, D: 8, dim1: 8, dim2: 9, dim3: 10, time: 20) != (dim1: 8, dim2: 9, dim3: 10, time: 20)
Coordinates only on the left object:
  * C        (C) float64 0.2328 0.9857 0.9737 0.4723 0.3088 -1.016 0.1256 -0.867
  * D        (D) float64 -0.378 -0.152 -0.9274 -0.6869 ... 0.7534 2.461 0.4231
Data variables only on the right object:
    C        (dim1) float64 0.2328 0.9857 0.9737 0.4723 ... -1.016 0.1256 -0.867
    D        (dim1) float64 -0.378 -0.152 -0.9274 ... 0.7534 2.461 0.4231
```

I don't think we need the numpy array test unless we're expecting it to align on dims. Are we?

Comment (max-sixty) [reply xarray/tests/test_dataset.py:3513]: (Though in the current code `test2` doesn't exist, so we need to change that to merge now)

**Comment by max-sixty** [comment]
@gcaria thanks a lot for the code and your patience working through the comments! We're very happy to have you as a contributor
